File: toolbox/+bct/+ui/js/console.js
/**
 * BCT Console JavaScript
 * 
 * Manages the console UI component including message display,
 * filtering, and command input.
 */

(function() {
    'use strict';
    
    const COMPONENT_ID = 'Console';
    let state = new BCT.StateManager({
        messages: [],
        maxMessages: 1000,
        autoScroll: true,
        showTimestamps: true,
        filterLevel: 0,
        initialized: false
    });
    
    // DOM elements
    let messagesContainer, inputField, filterSelect, btnClear, btnExport;
    
    /**
     * Initialize the console component
     */
    function init() {
        messagesContainer = document.getElementById('console-messages');
        inputField = document.getElementById('console-input');
        filterSelect = document.getElementById('filter-level');
        btnClear = document.getElementById('btn-clear');
        btnExport = document.getElementById('btn-export');
        
        if (!messagesContainer || !inputField) {
            console.error('Console elements not found');
            return;
        }
        
        // Set up event listeners
        inputField.addEventListener('keydown', handleInputKeydown);
        filterSelect.addEventListener('change', handleFilterChange);
        btnClear.addEventListener('click', clearMessages);
        btnExport.addEventListener('click', exportLog);
        
        // Set up MATLAB message listener
        setupMessageListener();
        
        // Notify MATLAB that component is ready
        BCT.sendToMatlab(COMPONENT_ID, 'ready');
        
        console.log('Console initialized');
    }
    
    /**
     * Set up listener for messages from MATLAB
     */
    function setupMessageListener() {
        BCT.onMatlabMessage(function(data) {
            if (data.id !== COMPONENT_ID) return;
            
            const cmd = data.cmd;
            
            switch (cmd) {
                case 'configure':
                    configure(data.config);
                    break;
                    
                case 'message':
                    addMessage(data.message);
                    break;
                    
                case 'clear':
                    clearMessages();
                    break;
                    
                case 'setFilter':
                    setFilter(data.level);
                    break;
                    
                case 'setMaxMessages':
                    state.set('maxMessages', data.max);
                    break;
                    
                case 'setAutoScroll':
                    state.set('autoScroll', data.enabled);
                    break;
                    
                case 'export':
                    exportToFile(data.filename);
                    break;
                    
                case 'enable':
                    enable();
                    break;
                    
                case 'disable':
                    disable();
                    break;
                    
                default:
                    console.warn('Unknown command:', cmd);
            }
        });
    }
    
    /**
     * Configure the console
     */
    function configure(config) {
        if (!config) {
            console.error('Invalid configuration');
            return;
        }
        
        state.set({
            maxMessages: config.maxMessages || 1000,
            autoScroll: config.autoScroll !== false,
            showTimestamps: config.showTimestamps !== false,
            filterLevel: config.filterLevel || 0,
            initialized: true
        });
        
        filterSelect.value = state.get('filterLevel');
    }
    
    /**
     * Add a message to the console
     */
    function addMessage(message) {
        const messages = state.get('messages');
        messages.push(message);
        
        // Enforce max messages limit
        const maxMessages = state.get('maxMessages');
        if (messages.length > maxMessages) {
            messages.shift();
        }
        
        state.set('messages', messages);
        
        // Render the message
        const messageElem = createMessageElement(message);
        messagesContainer.appendChild(messageElem);
        
        // Remove oldest if over limit
        const children = messagesContainer.children;
        if (children.length > maxMessages) {
            messagesContainer.removeChild(children[0]);
        }
        
        // Apply filter
        applyFilter();
        
        // Auto-scroll
        if (state.get('autoScroll')) {
            scrollToBottom();
        }
    }
    
    /**
     * Create a message element
     */
    function createMessageElement(message) {
        const elem = BCT.createElement('div', {
            classes: ['console-message', `console-message-${message.level}`],
            attrs: { 'data-id': message.id, 'data-level': getLevelValue(message.level) }
        });
        
        // Timestamp
        if (state.get('showTimestamps')) {
            const timestamp = BCT.createElement('span', {
                classes: 'console-timestamp',
                text: BCT.formatTime(message.timestamp * 1000)
            });
            elem.appendChild(timestamp);
        }
        
        // Level badge
        const badge = BCT.createElement('span', {
            classes: 'console-badge',
            text: getLevelBadge(message.level)
        });
        elem.appendChild(badge);
        
        // Message text
        const text = BCT.createElement('span', {
            classes: 'console-text',
            text: message.text
        });
        elem.appendChild(text);
        
        return elem;
    }
    
    /**
     * Get numeric level value
     */
    function getLevelValue(level) {
        const levels = {
            'info': 1,
            'success': 1,
            'warning': 2,
            'error': 3
        };
        return levels[level] || 0;
    }
    
    /**
     * Get level badge text
     */
    function getLevelBadge(level) {
        const badges = {
            'info': 'ℹ',
            'success': '✓',
            'warning': '⚠',
            'error': '✗'
        };
        return badges[level] || '·';
    }
    
    /**
     * Apply filter to messages
     */
    function applyFilter() {
        const filterLevel = state.get('filterLevel');
        const messages = messagesContainer.querySelectorAll('.console-message');
        
        messages.forEach(msg => {
            const msgLevel = parseInt(msg.getAttribute('data-level'));
            if (msgLevel >= filterLevel) {
                msg.style.display = '';
            } else {
                msg.style.display = 'none';
            }
        });
    }
    
    /**
     * Set filter level
     */
    function setFilter(level) {
        state.set('filterLevel', level);
        filterSelect.value = level;
        applyFilter();
    }
    
    /**
     * Handle filter dropdown change
     */
    function handleFilterChange() {
        setFilter(parseInt(filterSelect.value));
    }
    
    /**
     * Clear all messages
     */
    function clearMessages() {
        messagesContainer.innerHTML = '';
        state.set('messages', []);
    }
    
    /**
     * Handle input keydown (Enter to send command)
     */
    function handleInputKeydown(e) {
        if (e.key === 'Enter') {
            const input = inputField.value.trim();
            if (input) {
                sendCommand(input);
                inputField.value = '';
            }
        }
    }
    
    /**
     * Send command to MATLAB
     */
    function sendCommand(input) {
        BCT.sendToMatlab(COMPONENT_ID, 'command', { input: input });
        
        // Echo command in console
        addMessage({
            id: Date.now(),
            level: 'info',
            text: `> ${input}`,
            timestamp: Date.now() / 1000
        });
    }
    
    /**
     * Export log to file
     */
    function exportLog() {
        const messages = state.get('messages');
        
        if (messages.length === 0) {
            alert('No messages to export');
            return;
        }
        
        // Build export content
        const showTimestamps = state.get('showTimestamps');
        const lines = messages.map(msg => {
            let line = '';
            if (showTimestamps) {
                line += `[${BCT.formatTime(msg.timestamp * 1000)}] `;
            }
            line += `[${msg.level.toUpperCase()}] ${msg.text}`;
            return line;
        });
        
        const content = lines.join('\n');
        
        // Send to MATLAB for file writing
        BCT.sendToMatlab(COMPONENT_ID, 'exportData', {
            content: content,
            filename: 'console.txt'
        });
    }
    
    /**
     * Export to specific file
     */
    function exportToFile(filename) {
        const messages = state.get('messages');
        const showTimestamps = state.get('showTimestamps');
        
        const lines = messages.map(msg => {
            let line = '';
            if (showTimestamps) {
                line += `[${BCT.formatTime(msg.timestamp * 1000)}] `;
            }
            line += `[${msg.level.toUpperCase()}] ${msg.text}`;
            return line;
        });
        
        const content = lines.join('\n');
        
        BCT.sendToMatlab(COMPONENT_ID, 'exportData', {
            content: content,
            filename: filename
        });
    }
    
    /**
     * Scroll to bottom of messages
     */
    function scrollToBottom() {
        messagesContainer.scrollTop = messagesContainer.scrollHeight;
    }
    
    /**
     * Enable console
     */
    function enable() {
        inputField.disabled = false;
        filterSelect.disabled = false;
        btnClear.disabled = false;
        btnExport.disabled = false;
    }
    
    /**
     * Disable console
     */
    function disable() {
        inputField.disabled = true;
        filterSelect.disabled = true;
        btnClear.disabled = true;
        btnExport.disabled = true;
    }
    
    // Initialize when DOM is ready
    if (document.readyState === 'loading') {
        document.addEventListener('DOMContentLoaded', init);
    } else {
        init();
    }
    
})();

File: toolbox/+bct/+ui/Console.m
classdef Console < bct.ui.Component
    %CONSOLE HTML-based console component for BCT applications
    %
    %   The Console provides a logging and output display panel with
    %   support for different message types (info, warning, error),
    %   filtering, and command input.
    %
    %   Properties:
    %       MaxMessages     - Maximum number of messages to keep (default: 1000)
    %       AutoScroll      - Automatically scroll to bottom (default: true)
    %       ShowTimestamps  - Show timestamps for messages (default: true)
    %       FilterLevel     - Minimum message level to display
    %
    %   Methods:
    %       log        - Log an info message
    %       warn       - Log a warning message
    %       error      - Log an error message
    %       success    - Log a success message
    %       clear      - Clear all messages
    %       setFilter  - Set message filter level
    %       export     - Export console contents
    %
    %   Example:
    %       % In App Designer:
    %       app.Console = bct.ui.Console(app, app.HTMLConsole);
    %
    %       % Log messages:
    %       app.Console.log('Mesh loaded successfully');
    %       app.Console.warn('Eigenbasis computation may take several minutes');
    %       app.Console.error('Invalid signal dimensions');
    %
    %   See also: bct.ui.Component, bct.ui.Toolstrip
    
    properties (Access = public)
        MaxMessages     % Maximum messages to keep
        AutoScroll      % Auto-scroll to bottom
        ShowTimestamps  % Show message timestamps
        FilterLevel     % Message filter level (0=all, 1=info, 2=warn, 3=error)
    end
    
    properties (Access = private)
        MessageCount    % Total number of messages logged
    end
    
    methods
        function obj = Console(app, htmlComponent)
            %CONSOLE Construct a Console instance
            %
            %   obj = Console(app, htmlComponent) creates a console
            %   component and loads its HTML interface.
            
            % Call superclass constructor
            obj@bct.ui.Component(app, htmlComponent, "Console");
            
            % Initialize properties
            obj.MaxMessages = 1000;
            obj.AutoScroll = true;
            obj.ShowTimestamps = true;
            obj.FilterLevel = 0;  % Show all by default
            obj.MessageCount = 0;
            
            % Load the HTML interface
            obj.loadHTML("console.html");
            
            % Send initial configuration
            obj.sendConfiguration();
        end
        
        function log(obj, message, varargin)
            %LOG Log an info message
            %
            %   log(obj, message) logs an informational message.
            %   log(obj, format, args...) logs a formatted message.
            
            if nargin > 2
                message = sprintf(message, varargin{:});
            end
            obj.sendMessage('info', message);
        end
        
        function warn(obj, message, varargin)
            %WARN Log a warning message
            %
            %   warn(obj, message) logs a warning message.
            
            if nargin > 2
                message = sprintf(message, varargin{:});
            end
            obj.sendMessage('warning', message);
        end
        
        function error(obj, message, varargin)
            %ERROR Log an error message
            %
            %   error(obj, message) logs an error message.
            
            if nargin > 2
                message = sprintf(message, varargin{:});
            end
            obj.sendMessage('error', message);
        end
        
        function success(obj, message, varargin)
            %SUCCESS Log a success message
            %
            %   success(obj, message) logs a success message.
            
            if nargin > 2
                message = sprintf(message, varargin{:});
            end
            obj.sendMessage('success', message);
        end
        
        function clear(obj)
            %CLEAR Clear all messages from the console
            
            obj.MessageCount = 0;
            obj.send(struct('cmd', 'clear'));
        end
        
        function setFilter(obj, level)
            %SETFILTER Set message filter level
            %
            %   setFilter(obj, level) sets the minimum message level to
            %   display. Levels: 0=all, 1=info+, 2=warn+, 3=error only
            
            obj.FilterLevel = level;
            obj.send(struct('cmd', 'setFilter', 'level', level));
        end
        
        function exportLog(obj, filename)
            %EXPORTLOG Export console contents to file
            %
            %   exportLog(obj, filename) saves console messages to a file.
            
            if nargin < 2
                [file, path] = uiputfile('*.txt', 'Save Console Log');
                if isequal(file, 0)
                    return;
                end
                filename = fullfile(path, file);
            end
            
            % Request export from JavaScript
            obj.send(struct('cmd', 'export', 'filename', filename));
        end
        
        function setMaxMessages(obj, maxMessages)
            %SETMAXMESSAGES Set maximum number of messages to keep
            
            obj.MaxMessages = maxMessages;
            obj.send(struct('cmd', 'setMaxMessages', 'max', maxMessages));
        end
        
        function setAutoScroll(obj, enabled)
            %SETAUTOSCROLL Enable/disable auto-scrolling
            
            obj.AutoScroll = enabled;
            obj.send(struct('cmd', 'setAutoScroll', 'enabled', enabled));
        end
        
        function onMessage(obj, data)
            %ONMESSAGE Handle messages from JavaScript
            
            if ~isfield(data, 'cmd')
                return;
            end
            
            cmd = data.cmd;
            
            switch cmd
                case "command"
                    obj.handleCommand(data.input);
                    
                case "exportData"
                    obj.handleExport(data.content, data.filename);
                    
                case "ready"
                    obj.sendConfiguration();
                    
                otherwise
                    warning('bct:ui:Console:UnknownCommand', ...
                        'Unknown command: %s', cmd);
            end
        end
    end
    
    methods (Access = private)
        function sendConfiguration(obj)
            %SENDCONFIGURATION Send configuration to JavaScript
            
            config = struct(...
                'maxMessages', obj.MaxMessages, ...
                'autoScroll', obj.AutoScroll, ...
                'showTimestamps', obj.ShowTimestamps, ...
                'filterLevel', obj.FilterLevel);
            
            obj.send(struct('cmd', 'configure', 'config', config));
        end
        
        function sendMessage(obj, level, text)
            %SENDMESSAGE Send a log message to JavaScript
            
            obj.MessageCount = obj.MessageCount + 1;
            
            msg = struct(...
                'id', obj.MessageCount, ...
                'level', level, ...
                'text', text, ...
                'timestamp', posixtime(datetime('now')));
            
            obj.send(struct('cmd', 'message', 'message', msg));
        end
        
        function handleCommand(obj, input)
            %HANDLECOMMAND Handle command input from console
            
            % Route to app if it has a command handler
            if ismethod(obj.App, 'handleConsoleCommand')
                obj.App.handleConsoleCommand(input);
            else
                % Try to evaluate as MATLAB code (use with caution!)
                try
                    evalin('base', input);
                    obj.success('Command executed');
                catch ME
                    obj.error('Command failed: %s', ME.message);
                end
            end
        end
        
        function handleExport(obj, content, filename)
            %HANDLEEXPORT Handle export request from JavaScript
            
            try
                fid = fopen(filename, 'w');
                if fid == -1
                    error('Could not open file for writing');
                end
                
                fprintf(fid, '%s', content);
                fclose(fid);
                
                obj.success('Log exported to: %s', filename);
            catch ME
                obj.error('Export failed: %s', ME.message);
            end
        end
    end
end
